decisions/20240313-try-w3up.md

```diff
@@ -0,0 +1,142 @@
+# nft.storage should try w3up for web3 storage of new uploads
+
+Authors
+- bengo@protocol.ai
+
+## Problem Statement
+
+nft.storage needs to offer web3 storage and currently does so using a homegrown solution that may be expensive to operate, maintain, and scale.
+
+Meanwhile the https://web3.storage team has productized a scalable web3 storage solution (named [w3up](https://github.com/web3-storage/w3up/)), and we think we may be able to deliver the same functionality that nft.storage needs for cheaper than it costs nft.storage to operate and improve the homegrown solution they use today that relies on [dagcargo](https://github.com/nftstorage/dagcargo).
+
+But it's not clear how much it would cost for nft.storage to adopt web3.storage and interoperate with w3up.
+
+If we lower the cost of adoption enough, perhaps nft.storage would be happy to ship the integration and use w3up/web3.storage as their storage/ipfs-serving/filecoin-dealmaking partner for the foreseeable future.
+
+If we scout out the integration a bit, in the best case we'll enable short-term adoption of the product and, no matter what, we'll learn more about how usable our tools are in this real-world integration style, and we can use that to inform product improvements that hopefully help adoption of web3.storage in medium term.
+
+## Goals
+
+* enable nft.storage to store new uploads using web3.storage
+* enable nft.storage to migrate to web3.storage, have an account there we have in our CRM, access console.web3.storage, etc
+
+## Non-Goals
+
+* migrate old nft.storage uploads
+* fully remove reliance on dagcargo
+  * however, this work may serve as useful input to that. If nft.storage can upload using w3up, then w3up can probably handle the filecoin aggregation via [w3-filecoin](https://github.com/web3-storage/specs/blob/main/w3-filecoin.md). And if we have that, it reduces reliance on dagcargo.
```

Comment:
@gobengo do you mean by this that first iteration is behind a feature flag via ENV VAR + specific accounts? Or do you mean for first iteration we still need dagcargo DB, but not the aggregator given all uploads will go via w3up, but we still need DB until second stage?

Reply:
> do you mean by this that first iteration is behind a feature flag via ENV VAR + specific accounts?

I mostly meant this. 'fully remove reliance on dagcargo' is a nongoal for first iteration of just making sure we can do at least some uploads via w3up. But I wrote this bullet to say that, while it's a non-goal this sprint, I hope this moves us closer to removing dagcargo by replacing it with w3up.

Assuming this scouting goal goes well, 'fully remove reliance on dagcargo' could be next goal.

Reply:
For me that is fine if the date dagcargo goes off is before sprint ends? worth double checking with @reidlw

Reply:
> For me that is fine if the date dagcargo goes off is before sprint ends? 

I think the sprint ends on March 27, and my last day of availability is March 22. I'll be back April 1. I don't see a path to dagcargo going off before the sprint ends.